### PR TITLE
Handle errors when chart is invalid for given data

### LINKED ISSUE #1150: Various types of query charts throw exceptions and don't render
- SQL Operations Studio Version: 0.28.3

Steps to Reproduce:
Trying to repro https://github.com/Microsoft/sqlopsstudio/issues/32 I hit the below exception

1. Run this query:
`select '2017-06-22T00:00:00' [dt], 791 [count] 
union all select '2017-06-22T01:00:00' [dt], 451
union all select '2017-06-22T02:00:00' [dt], 528
`

2. Click the "View as Chart" button in the bottom Results pane, on the right side.
3. In the Chart Type drop down menu, choose "timeSeries" *(exception happens here)*

```
ERR Cannot read property 'data' of undefined: TypeError: Cannot read property 'data' of undefined
    at BaseChartDirective.updateChartData (D:\xplat\sqlopsstudio\node_modules\ng2-charts\charts\charts.js:88:43)
    at BaseChartDirective.ngOnChanges (D:\xplat\sqlopsstudio\node_modules\ng2-charts\charts\charts.js:39:26)
    at checkAndUpdateDirectiveInline (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:10819:19)
    at checkAndUpdateNodeInline (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12245:17)
    at checkAndUpdateNode (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12184:16)
    at debugCheckAndUpdateNode (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12887:59)
    at debugCheckDirectivesFn (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12828:13)
    at Object.eval [as updateDirectives] (ng:///QueryOutputModule/TimeSeriesChart.ngfactory.js:61:5)
    at Object.debugUpdateDirectives [as updateDirectives] (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12813:21)
    at checkAndUpdateView (D:\xplat\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12151:14)
```

### PR DESCRIPTION
**Pull request by MattIrv**
Fixes #1150 

Some chart styles fail to display when the query results for those charts don't match the expected format, causing the entire chart viewer to break after selecting one of the invalid charts. This change handles those errors more gracefully by catching the error and displaying an error message to the user.

This change also disables the "Image" type in chart viewer since it throws errors related to `_config` being missing and there's currently no way for the user to set `_config` from the chart viewer

### CODE REVIEW

**Comment by anthonydresser** [comment]
I don't think the `validasquery` is the right solution. People have already logged bugs against the image chart viewer, meaning they are using it. So I don't think it's correct to just remove it.

File: src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts [hunk 1]
```diff
@@ -129,7 +134,14 @@ export abstract class ChartInsight extends Disposable implements IInsightsView {
 		// hence it's easier to not render until ready
 		this.options = mixin(this.options, { maintainAspectRatio: false });
 		this._hasInit = true;
-		this._changeRef.detectChanges();
+		this._hasError = false;
+		try {
+			this._changeRef.detectChanges();
```
Comment (anthonydresser) [inline src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts]: What is the instance when this errors? Generally if this is happening there is greater angular logic we are not doing correctly. So just wrapping it in a try catch might fix this narrow problem, but there will probably be other related problems.

**Comment by MattIrv** [comment]
@anthonydresser I didn't see a way that the image viewer could ever work as a chart type in chart viewer given that we don't expose those configs right now. Do you have an example of how to use it so I can take a look? Or was there some recent regression that broke it?

File: src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts (hunk 1 above)
Comment (MattIrv) [reply src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts]: The error comes from within ng2-charts code based on trying to use data that isn't in the expected format. Unless we want to manually verify the data format for each chart type (might be a good idea but would take more time - do you know where that format is documented?) then I'm not sure how else we'd catch it.

Here's an example stack trace:
```
/Users/mairvine/code/sqlopsstudio/node_modules/zone.js/dist/zone-node.js:2251   ERR ng-charts configuration error,
      data or datasets field are required to render char scatter: Error: ng-charts configuration error,
      data or datasets field are required to render char scatter
    at BaseChartDirective.getDatasets (/Users/mairvine/code/sqlopsstudio/node_modules/ng2-charts/charts/charts.js:129:19)
    at BaseChartDirective.getChartBuilder (/Users/mairvine/code/sqlopsstudio/node_modules/ng2-charts/charts/charts.js:57:29)
    at BaseChartDirective.refresh (/Users/mairvine/code/sqlopsstudio/node_modules/ng2-charts/charts/charts.js:139:27)
    at BaseChartDirective.ngOnInit (/Users/mairvine/code/sqlopsstudio/node_modules/ng2-charts/charts/charts.js:28:18)
    at checkAndUpdateDirectiveInline (/Users/mairvine/code/sqlopsstudio/node_modules/@angular/core/bundles/core.umd.js:10822:19)
    at checkAndUpdateNodeInline (/Users/mairvine/code/sqlopsstudio/node_modules/@angular/core/bundles/core.umd.js:12245:17)
    at checkAndUpdateNode (/Users/mairvine/code/sqlopsstudio/node_modules/@angular/core/bundles/core.umd.js:12184:16)
    at debugCheckAndUpdateNode (/Users/mairvine/code/sqlopsstudio/node_modules/@angular/core/bundles/core.umd.js:12887:59)
    at debugCheckDirectivesFn (/Users/mairvine/code/sqlopsstudio/node_modules/@angular/core/bundles/core.umd.js:12828:13)
    at Object.eval [as updateDirectives] (ng:///QueryOutputModule/ScatterChart.ngfactory.js:61:5)
```

**Comment by anthonydresser** [comment]
It looks like the problem is that https://github.com/Microsoft/sqlopsstudio/blob/master/src/sql/parts/dashboard/widgets/insights/views/imageInsight.component.ts#L65 is accessing `this._config` without checking to make sure it is defined. So I think the proper fix here is to just add a check to make sure `this._config` is defined. The 'config' for image insight is rather limited. So a image insight can perfectly work as an insight assuming the first cell/row is a base64 image encoding. The eventually perfect fix would be to implement the ui to control the config for the image insight, just as we do for all other chart types.

**Comment by kburtram** [comment]
Do we know if #1150 is a regression from March Public Preview?  ...I'm trying to determine if this is something we need to take today, or if it can wait until May.

**Comment by anthonydresser** [comment]
It's definitely not a regression. It has always worked this way, but no one put in the effort to fix all the different use cases.

**Comment by MattIrv** [comment]
@anthonydresser thanks, just updated the image insight to use the default config and removed the `validasquerychart` logic.

**Comment by kburtram** [comment]
Thanks @anthonydresser!  @MattIrv if this isn't a regression we don't need it for today.  But if you merge before 7pm I'll pick it up when I fork the release branch.

File: src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts (hunk 1 above)
Comment (anthonydresser) [reply src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts]: All the different charts are here http://www.chartjs.org/docs/latest/charts/ we probably should verify all the data and try to be smart about letting the user know why we can't render their chart. 

I attempted to add validation here https://github.com/Microsoft/sqlopsstudio/blob/master/src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts#L326 but clearly I didn't catch everything.

Comment (MattIrv) [reply src/sql/parts/dashboard/widgets/insights/views/charts/chartInsight.component.ts]: @anthonydresser that's probably a bigger change than we want to take on right before the release. Would you be ok with merging this change as-is and then we can make those improvements later? As it stands this is a good user experience improvement for this release

**Pull request merged**